substrate/frame/revive/src/vm/evm/instructions/host.rs

```diff
@@ -15,146 +15,119 @@
-		gas_legacy!(context.interpreter, warm_cold_cost(code.is_cold));
-	} else if spec_id.is_enabled_in(TANGERINE) {
-		gas_legacy!(context.interpreter, 700);
-	} else {
-		gas_legacy!(context.interpreter, 20);
-	}
-
-	*top = U256::from(code.len());
+	gas!(context.interpreter, RuntimeCosts::CodeSize);
+	let h160 = sp_core::H160::from_slice(&top.to_be_bytes::<32>()[12..]);
+	let code_size = context.interpreter.extend.code_size(&h160);
+	*top = U256::from(code_size);
 }
 
 /// EIP-1052: EXTCODEHASH opcode
 pub fn extcodehash<'ext, E: Ext>(context: Context<'_, 'ext, E>) {
-	check!(context.interpreter, CONSTANTINOPLE);
 	popn_top!([], top, context.interpreter);
-	let address = top.into_address();
-	let Some(code_hash) = context.host.load_account_code_hash(address) else {
-		context.interpreter.halt(InstructionResult::FatalExternalError);
-		return;
-	};
-	let spec_id = context.interpreter.runtime_flag.spec_id();
-	if spec_id.is_enabled_in(BERLIN) {
-		gas_legacy!(context.interpreter, warm_cold_cost(code_hash.is_cold));
-	} else if spec_id.is_enabled_in(ISTANBUL) {
-		gas_legacy!(context.interpreter, 700);
-	} else {
-		gas_legacy!(context.interpreter, 400);
-	}
-	*top = code_hash.into_u256();
+	gas!(context.interpreter, RuntimeCosts::CodeHash);
+	let h160 = sp_core::H160::from_slice(&top.to_be_bytes::<32>()[12..]);
+	let code_hash = context.interpreter.extend.code_hash(&h160);
+	*top = U256::from_be_bytes(code_hash.0);
 }
 
 /// Implements the EXTCODECOPY instruction.
 ///
 /// Copies a portion of an account's code to memory.
 pub fn extcodecopy<'ext, E: Ext>(context: Context<'_, 'ext, E>) {
 	popn!([address, memory_offset, code_offset, len_u256], context.interpreter);
-	let address = address.into_address();
-	let Some(code) = context.host.load_account_code(address) else {
+
+	let h160 = sp_core::H160::from_slice(&address.to_be_bytes::<32>()[12..]);
+	let code_hash = context.interpreter.extend.code_hash(&h160);
+
+	let Some(code) =
+		crate::PristineCode::<E::T>::get(&code_hash).map(|bounded_vec| bounded_vec.to_vec())
+	else {
 		context.interpreter.halt(InstructionResult::FatalExternalError);
 		return;
 	};
-
-	let len = as_usize_or_fail!(context.interpreter, len_u256);
-	gas_or_fail_legacy!(
-		context.interpreter,
-		gas::extcodecopy_cost(context.interpreter.runtime_flag.spec_id(), len, code.is_cold)
-	);
-	if len == 0 {
+	let Ok(code_offset) = code_offset.try_into() else {
+		context.interpreter.halt(InstructionResult::Revert);
 		return;
-	}
-	let memory_offset = as_usize_or_fail!(context.interpreter, memory_offset);
-	let code_offset = min(as_usize_saturated!(code_offset), code.len());
-	resize_memory!(context.interpreter, memory_offset, len);
+	};
+	if code_offset >= code.len() {
+		context.interpreter.halt(InstructionResult::Revert);
+		return;
+	};
+	let Ok(memory_len): Result<usize, _> = len_u256.try_into() else {
+		context.interpreter.halt(InstructionResult::Revert);
+		return;
+	};
+	if memory_len > code.len().saturating_sub(code_offset) {
```

Comment:
Please mind PGs comment above. The semantics are incorrect. Instead `EXTCODECOPY` zeroes out OOB bytes.